prov/efa/src/rxr/rxr_ep.c

```diff
@@ -2484,6 +2486,59 @@ int rxr_ep_alloc_app_device_info(struct fi_info **app_device_info,
 	return 0;
 }
 
+/**
+ * @brief Create ibv_cq_ex by calling ibv_create_cq_ex
+ *
+ * @param[in] cq Pointer to the efa_cq.
+ * @param[in] attr Pointer to fi_cq_attr.
+ * @param[out] Return pointer to ibv_cq_ex if successful, otherwise NULL
+ */
+static inline struct ibv_cq_ex *rxr_ep_create_ibv_cq_ex(struct rxr_ep *ep, struct fi_cq_attr *attr)
+{
+	struct ibv_cq_init_attr_ex init_attr_ex = {
+		.cqe = attr->size ? attr->size : EFA_DEF_CQ_SIZE,
+		.cq_context = NULL,
+		.channel = NULL,
+		.comp_vector = 0,
+		/* EFA requires these values for wc_flags and comp_mask.
+		 * See `efa_create_cq_ex` in rdma-core.
+		 */
+		.wc_flags = IBV_WC_STANDARD_FLAGS,
+		.comp_mask = 0,
+	};
+
+	struct efa_domain *efa_domain = rxr_ep_domain(ep);
+
+	return ibv_create_cq_ex(efa_domain->device->ibv_ctx, &init_attr_ex);
+}
+
+/**
+ * @brief Bind rxr_ep->ibv_cq_ex to efa_ep->scq and efa_ep->rcq
+ * 	Will be removed when rxr_ep and efa_ep are fully separated
+ *
+ * @param[in] rxr_ep Pointer to the rxr_ep.
+ * @param[in] ibv_cq_ex Pointer to ibv_cq_ex.
+ */
+static int rxr_ep_setup_device_cq(struct rxr_ep *rxr_ep)
+{
+	struct efa_ep *efa_ep = container_of(rxr_ep->rdm_ep, struct efa_ep, util_ep.ep_fid);
+	struct ibv_cq_ex *ibv_cq_ex = rxr_ep->ibv_cq_ex;
+
+	efa_ep->scq = malloc(sizeof(struct efa_cq));
+	if (!efa_ep->scq)
+		return -FI_ENOMEM;
+	efa_ep->scq->ibv_cq_ex = ibv_cq_ex;
+	efa_ep->scq->domain = rxr_ep_domain(rxr_ep);
+
+	efa_ep->rcq = malloc(sizeof(struct efa_cq));
+	if (!efa_ep->rcq)
```

Comment:
when rcq malloc failed, do you need to free the `scq` ?

Reply:
I think you do not need to allocate `rcq` here. set `efa_ep->rcq` to `efa_ep->scq` should work.

Reply:
fixed